cassandra/connection.py

```diff
@@ -1188,11 +1207,22 @@ def process_msg(self, header, body):
                 decoder = paging_session.decoder
                 result_metadata = None
             else:
+                need_notify_of_release = False
+                with self.lock:
+                    if stream_id in self.orphaned_request_ids:
+                        self.in_flight -= 1
+                        self.orphaned_request_ids.remove(stream_id)
+                        need_notify_of_release = True
+                if need_notify_of_release and self._on_orphaned_stream_released:
+                    self._on_orphaned_stream_released()
+
                 try:
                     callback, decoder, result_metadata = self._requests.pop(stream_id)
                 # This can only happen if the stream_id was
                 # removed due to an OperationTimedOut
                 except KeyError:
+                    with self.lock:
+                        self.request_ids.append(stream_id)
```

Comment:
Just to be very clear... this change isn't explicitly connected to the problem you're trying to address in the PR, correct @haaawk ?  The intent here is that if we get a stream ID that doesn't match up to a request we know about we should be safe to re-use that stream ID regardless of other conditions... do I have that right?

Reply:
In https://github.com/datastax/python-driver/pull/1114/commits/12c9c306d39989142b21d06a315f6657d42d1eef:
1. If a request R has timed out due to a client side timeout, we stopped returning its stream ID to the pool of available stream IDs
2. If the same request R receives a response after R has timed out on a client side, the stream ID of R will be put back to the pool of available stream IDs

This line performs (2).

>if we get a stream ID that doesn't match up to a request we know about we should be safe to re-use that stream ID regardless of other conditions... do I have that right?

Yes. The assumption is that if we got a stream ID we don't know then it must be a stream ID of a request that has already timed out due to a client side timeout.

Reply:
Okay, thanks for the verification @haaawk !